### ISSUE TITLE
Some MPS functionality regressed in MacOS-14.4

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#125803]
### 🐛 Describe the bug

For example following works on 13.x and 14.2, but fails on 14.4:
```
% python3 test_mps.py -k TestRNNMPS.test_lstm_backward
F
======================================================================
FAIL: test_lstm_backward (__main__.TestRNNMPS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nshulga/git/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 2763, in wrapper
    method(*args, **kwargs)
  File "/Users/nshulga/git/pytorch/pytorch/test/test_mps.py", line 11420, in test_lstm_backward
    self._lstm_helper(num_layers=num_layers, dtype=dtype, device=device, backward=True, **test_options)
  File "/Users/nshulga/git/pytorch/pytorch/test/test_mps.py", line 11398, in _lstm_helper
    self.assertEqual(cpu_weight_grad, mps_weight_grad,
  File "/Users/nshulga/git/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 3646, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 195 / 196 (99.5%)
Greatest absolute difference: 0.24748152494430542 at index (17, 1) (up to 1e-05 allowed)
Greatest relative difference: 59.33826446533203 at index (10, 2) (up to 1.3e-06 allowed)
mismatch in cpu:weight_ih_l2 vs mps:weight_ih_l2, layers: 5

To execute this test, run the following from the base repo dir:
     python test/test_mps.py -k TestRNNMPS.test_lstm_backward

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 1 test in 1.186s

FAILED (failures=1)

```

### Versions

MacOS-14.4

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @seemethere @pytorch/pytorch-dev-infra @kulinseth @albanD @DenisVieriu97 @jhavukainen

**Comment by shink** [comment]
22 tests failed:

![](https://github.com/pytorch/pytorch/assets/49934421/70352965-fcfd-441d-9d3d-703595cf7426)

versions:

```bash
$ python torch/utils/collect_env.py
Collecting environment information...
PyTorch version: 2.4.0a0+gitfc2256f
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.11.9 (main, Apr 19 2024, 11:43:47) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.9.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] optree==0.11.0
[pip3] torch==2.4.0a0+gitfc2256f
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] torch                     2.4.0a0+gitfc2256f          pypi_0    pypi
[conda] torchfix                  0.4.0                    pypi_0    pypi
```

**Comment by kulinseth** [comment]
`lstm` RNN tests did regress. @jhavukainen is taking a look at it...

**Comment by jhavukainen** [comment]
We have found the root cause for the issue to a bug in an optimization that went into MPSGraph reshape operation in MacOS 14.4 which only triggers here once the num_layers > 2. The fix will require a work on the kernel side so it'll be tied to a future OS release. In the meanwhile I'll add an assert to instruct the user to fallback to CPU if they want to use num_layers > 2:

https://github.com/pytorch/pytorch/pull/130038

**Comment by jhavukainen** [comment]
The issue was resolved with #130038
